httpbakery/form/form.go

```diff
@@ -0,0 +1,151 @@
+	      |                                    |
+	      | GET "form" URL                     |
+	      |----------------------------------->|
+	      |                                    |
+	      |                  Schema definition |
+	      |<-----------------------------------|
+	      |                                    |
+	+-------------+                            |
+	|   Client    |                            |
+	| Interaction |                            |
+	+-------------+                            |
+	      |                                    |
+	      | POST data to "form" URL            |
+	      |----------------------------------->|
+	      |                                    |
+	      |                Form login response |
+	      |<-----------------------------------|
+	      |                                    |
+
+The schema is provided as a environschema.Fileds object. It is the
+client's responsibility to interpret the schema and present it to the
+user.
+*/
+
+// Objects implementing the FormProcessor interface interpret schema and
+// interact with the user. When the interaction is complete the returned
+// value should conform to the provided schema.
+type Processor interface {
+	ProcessForm(schema environschema.Fields) (interface{}, error)
+}
+
+// SetUpAuth configures form authentication on c. The VisitWebPage field
+// in c will be set to a function that will attempt form based
+// authentication using h to perform the interaction with the user.
+//
+// The VisitWebPage function downloads the schema from the specified
+// server and calls h.Handle. The map returned by h.Handle should match
+// the schema specified, but VisitWebPage will not verify this before
+// sending to the server. Any errors returned by h.Handle will not have
+// their cause masked.
+func SetUpAuth(c *httpbakery.Client, p Processor) {
+	v := webPageVisitor{
+		client: &httprequest.Client{
+			Doer: c,
+		},
+		processor: p,
+	}
+	c.VisitWebPage = v.visitWebPage
+}
+
+// webPageVisitor contains the state required by visitWebPage.
+type webPageVisitor struct {
+	client    *httprequest.Client
+	processor Processor
+}
+
+// loginMethods contains the response expected from the login URL. It
+// only checks for the "form" method as that is the only one that can be
+// handled.
+type loginMethods struct {
+	Form string `json:"form"`
+}
+
+// SchemaRequest is a request for a form schema.
+type SchemaRequest struct {
+	httprequest.Route `httprequest:"GET"`
+}
+
+// SchemaResponse contains the message expected in response to the schema
+// request.
+type SchemaResponse struct {
+	Schema environschema.Fields `json:"schema"`
+}
+
+// LoginRequest is a request to perform a login using the provided details.
+type LoginRequest struct {
+	httprequest.Route `httprequest:"POST"`
+	Login             LoginWrapper `httprequest:",body"`
+}
+
+// Login holds the body of a login request.
+type LoginWrapper struct {
+	Login interface{} `json:"login"`
```

Comment:
This one specifically I don't want to be a map, if it is an interface I can do creative things on servers with real types.